### ISSUE TITLE
[help wanted] : Problem when training/using a POS tagger with Flair

### ISSUE THREAD

**Issue by lkevers** [flairNLP/flair#3194]
I have a problem when training a POS tagger with Flair (Version: 0.11.3), or more precisely when using the resulting learned model.

The training data is in CONLLU format. For the sentence "Allora ùn ti dicu nulla, ùn ti dicu nulla !", the CONLLU is the following :
```
# text = Allora ùn ti dicu nulla, ùn ti dicu nulla !
1       Allora  _       ADV     _       _       0       _       _       _
2       ùn      _       ADV     _       _       1       _       _       _
3       ti      _       PRON    _       _       1       _       _       _
4       dicu    _       VERB    _       _       1       _       _       _
5       nulla   _       PRON    _       _       1       _       _       SpaceAfter=No
6       ,       _       PUNCT   _       _       1       _       _       _
7       ùn      _       ADV     _       _       1       _       _       _
8       ti      _       PRON    _       _       1       _       _       _
9       dicu    _       VERB    _       _       1       _       _       _
10      nulla   _       PRON    _       _       1       _       _       _
11      !       _       PUNCT   _       _       1       _       _       _
```
I currently have very little learning data. I have done these first Flair tests with 100 sentences.

The training is done as described below from CONNLU files.

**1. TRAIN the POS tagger**

```
# a. Load data (CONLLU format)
columns = {0: 'id', 1: 'form', 2: 'lemma', 3: 'upos', 4: 'xpos', 5: 'feats', 6: 'head', 7:'deprel', 8:'deps', 9:'misc'}
corpus_cos: Corpus = ColumnCorpus(foldPath, columns,
                                        train_file='TRAIN.conllu',
                                        test_file='TEST.conllu',
                                        dev_file='DEV.conllu')
label_type = 'upos'
label_dict = corpus.make_label_dictionary(label_type=label_type)

# b. embeddings
embeddings = StackedEmbeddings(embeddings=FlairEmbeddings(it-forward))

# c. define tagger
tagger = SequenceTagger(hidden_size=hiddenSize,
                            embeddings=embeddings,
                            tag_dictionary=label_dict,
                            tag_type=label_type,
                            use_crf=True)
# d. define trainer
trainer = ModelTrainer(tagger, corpus)

# e. learn
foldOutputDir="%s/%s"%(outputDir,str(fold).rjust(2,'0'))
trainer.train(foldOutputDir,
                learning_rate=trainLearningRate,
                mini_batch_size=trainMiniBatchSize,
                max_epochs=trainMaxEpochs)
```
                
In order to get the details of the tags, I re-run the evaluation with the out_path argument as below:

```
result = tagger.evaluate(corpus_cos.test, mini_batch_size=32, out_path=f"%s/predictionsWithBestModel.txt"%foldOutputDir, gold_label_type="upos")
print(result.detailed_results)
```

I then use the best learned model to tag the TEST corpus (identical to the one used during training). The data to be annotated is loaded into a list of Sentences, each Sentence being a string containing the tokens separated by a space (this is why I used the mention use_tokenizer=False).

**2. PREDICT pos tags on TEST data using the learned model**

```
model = SequenceTagger.load('%s/best-model.pt'%foldOutputDir)
goldFile="%s/TEST.conllu"%foldPath
gold=pyconll.load_from_file(goldFile)

txtTag=[]
    for sentGold in gold :
        sent=""
        for token in sentGold :
            sent="%s %s"%(sent,token.form)
        txtTag.append(Sentence(sent.strip(),use_tokenizer=False)) # use_tokenizer=False : no tokenization is performed and the text is split on whitespaces

model.predict(txtTag)
```

The results obtained at the end of the training (accuracy=0.2634) are very different from those I obtain using the best model (accuracy=0.1452).

The comparison of the results obtained shows that we do not obtain the same tags (differences are prefixed by *). 

**-> Evaluation after training on CONLLU data :**
ID, GOLD, TAG
```
1 ADV DET                                                                                                
2 ADV NOUN                                                                                               
3 PRON ADP                                                                                               
4 VERB NOUN                                                                                              
5 PRON ADP                                                                                               
*6 PUNCT DET                                                                                              
*7 ADV NOUN                                                                                               
*8 PRON ADP                                                                                               
*9 VERB DET                                                                                               
*10 PRON NOUN                                                                                             
*11 PUNCT PUNCT
```

**-> Tagging with best model (on a string containing the tokens separated by spaces) :**
```
[0.1 - 0] Allora - GOLD : ADV <-> TAG NOTOK : DET
[0.2 - 1] ùn - GOLD : ADV <-> TAG NOTOK : NOUN 
[0.3 - 2] ti - GOLD : PRON <-> TAG NOTOK : ADP 
[0.4 - 3] dicu - GOLD : VERB <-> TAG NOTOK : NOUN
[0.5 - 4] nulla - GOLD : PRON <-> TAG NOTOK : ADP
*[0.6 - 5] , - GOLD : PUNCT <-> TAG NOTOK : NOUN
*[0.7 - 6] ùn - GOLD : ADV <-> TAG NOTOK : ADP
*[0.8 - 7] ti - GOLD : PRON <-> TAG NOTOK : DET
*[0.9 - 8] dicu - GOLD : VERB <-> TAG NOTOK : NOUN
*[0.10 - 9] nulla - GOLD : PRON <-> TAG NOTOK : ADP
*[0.11 - 10] ! - GOLD : PUNCT <-> TAG NOTOK : INTJ
```

Is the way I submit the text to be tagged correct? Where could this difference in tagging come from?

**Comment by lkevers** [comment]
I managed to solve this problem by transforming my CONLLU input format into a simplified tabular format in which unused columns do not appear. Practically, all I have left is the column containing the form and the column containing the POS tag. By loading this data everything works perfectly!

**Comment by alanakbik** [comment]
Hello @lkevers thanks for sharing the solution. I wonder if the issue is somehow caused by the CoNLL-U format. Have you tried printing the sentences from the ColumnCorpus to check if they are correctly read?

**Comment by lkevers** [comment]
Hi @alanakbik ! The output of the corpus two first sentences are:

- When loading a full CONLLU file : 

Sentence: "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18" → ["1"/S'/_/SCONJ/_/_/_, "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18", "2"/ellu/_/PRON/_/_/_, "3"/ci/_/PRON/_/_/_, "4"/hè/_/VERB/_/_/_, "5"/assai/_/DET/_/_/_, "6"/castagne/_/NOUN/_/_/_, "7"/digià/_/ADV/_/_/_, "8"/in/_/ADP/_/_/_, "9"/terra/_/NOUN/_/_/_, "10"/,/_/PUNCT/_/_/_, "11"/sonanu/_/VERB/_/_/_, "12"/u/_/DET/_/_/_, "13"/cornu/_/NOUN/_/_/_, "14"/è/_/CCONJ/_/_/_, "15"/dicenu/_/VERB/_/_/_, "16"/furestu/_/NOUN/_/_/_, "17"/serratu/_/ADJ/_/_/_, "18"/./_/PUNCT/_/_/_]

Sentence: "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24" → ["1"/M'/_/PRON/_/_/_, "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24", "2"/arricordu/_/VERB/_/_/_, "3"/chì/_/SCONJ/_/_/_, "4"/stu/_/DET/_/_/_, "5"/vecchju/_/ADJ/_/_/_, "6"/omu/_/NOUN/_/_/_, "7"/sprimunghjava/_/VERB/_/_/_, "8"/a/_/DET/_/_/_, "9"/cera/_/NOUN/_/_/_, "10"/pè/_/ADP/_/_/_, "11"/caccià/_/VERB/_/_/_, "12"/u/_/DET/_/_/_, "13"/mele/_/NOUN/_/_/_, "14"/,/_/PUNCT/_/_/_, "15"/certe/_/DET/_/_/_, "16"/volte/_/NOUN/_/_/_, "17"/facìa/_/VERB/_/_/_, "18"/bolle/_/NOUN/_/_/_, "19"/a/_/DET/_/_/_, "20"/cera/_/NOUN/_/_/_, "21"/pè/_/ADP/_/_/_, "22"/rèndela/_/VERB/_/_/_, "23"/liquida/_/ADJ/_/_/_, "24"/./_/PUNCT/_/_/_]

- When loading a simplified file (only two columns, "form" and "upos") : 

Sentence: "S' ellu ci hè assai castagne digià in terra , sonanu u cornu è dicenu furestu serratu ." → ["S'"/SCONJ, "ellu"/PRON, "ci"/PRON, "hè"/VERB, "assai"/ADV, "castagne"/NOUN, "digià"/ADV, "in"/ADP, "terra"/NOUN, ","/PUNCT, "sonanu"/VERB, "u"/DET, "cornu"/NOUN, "è"/CCONJ, "dicenu"/VERB, "furestu"/NOUN, "serratu"/ADJ, "."/PUNCT]

Sentence: "Ghjunghjìamu i piscioni da a furesta , a dera per scarlà u fornu ." → ["Ghjunghjìamu"/VERB, "i"/DET, "piscioni"/NOUN, "da"/ADP, "a"/DET, "furesta"/NOUN, ","/PUNCT, "a"/DET, "dera"/NOUN, "per"/ADP, "scarlà"/VERB, "u"/DET, "fornu"/NOUN, "."/PUNCT]

It seems to me that the problem could come from the first column which contains an index number of the tokens in the sentence.

**Comment by alanakbik** [comment]
Hello @lkevers yes that is probably the problem! You could try loading the original corpus with the following: 
```python
columns = {0: 'id', 1: 'text', 2: 'lemma', 3: 'upos', 4: 'xpos', 5: 'feats', 6: 'head', 7:'deprel', 8:'deps', 9:'misc'}
```
i.e. 'text' instead of 'form'. Then it should correctly recognize the second column as the words.
